web/src/settings_org.js

```diff
@@ -266,41 +266,19 @@ function message_delete_limit_setting_enabled() {
     // should be enabled. The setting is disabled when every user
     // who is allowed to delete their own messages is also allowed
     // to delete any message in the organization.
-    const realm_delete_own_message_policy = Number.parseInt(
-        $("#id_realm_delete_own_message_policy").val(),
-        10,
-    );
+    const realm_can_delete_own_message_group_id =
+        settings_components.get_dropdown_list_widget_setting_value(
+            $("#id_realm_can_delete_own_message_group"),
+        );
     const realm_can_delete_any_message_group_id =
         settings_components.get_dropdown_list_widget_setting_value(
             $("#id_realm_can_delete_any_message_group"),
         );
-    const realm_can_delete_any_message_group_name = user_groups.get_user_group_from_id(
-        realm_can_delete_any_message_group_id,
-    ).name;
-    const common_message_policy_values = settings_config.common_message_policy_values;
-
-    if (realm_delete_own_message_policy === common_message_policy_values.by_admins_only.code) {
-        return false;
-    }
-    if (realm_can_delete_any_message_group_name === "role:administrators") {
-        return true;
-    }
-    if (realm_delete_own_message_policy === common_message_policy_values.by_moderators_only.code) {
-        return false;
-    }
-    if (realm_can_delete_any_message_group_name === "role:moderators") {
-        return true;
-    }
-    if (realm_delete_own_message_policy === common_message_policy_values.by_full_members.code) {
-        return false;
-    }
-    if (realm_can_delete_any_message_group_name === "role:fullmembers") {
-        return true;
-    }
-    if (realm_delete_own_message_policy === common_message_policy_values.by_members.code) {
-        return false;
-    }
-    return true;
+    const can_delete_any_message_subgroups = user_groups.get_recursive_subgroups(
+        user_groups.get_user_group_from_id(realm_can_delete_any_message_group_id),
+    );
+    can_delete_any_message_subgroups.add(realm_can_delete_any_message_group_id);
+    return !can_delete_any_message_subgroups.has(realm_can_delete_own_message_group_id);
 }
```

Comment:
This logic will need to be reworked when we start permitting non-system-group values for these settings following https://github.com/zulip/zulip/pull/31527, but it's fine for now.